### ISSUE TITLE
[MNG-6720] MultiThreadedBuilder: wait for parallel running projects when using --fail-fast

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8531]
**[Stefan Oehme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oehme)** opened **[MNG-6720](https://issues.apache.org/jira/browse/MNG-6720?redirect=false)** and commented

When a project fails using --fail-fast mode, other projects that are currently running in parallel are not finished before the "session finished" event is sent. This leads to issues for extensions that rely on the documented order of events, which says that "session finished" is only sent "after all projects have been finished/skipped".

In our case, we close a service on "session finished", and that service is needed to execute goals. The goals that were running in parallel at that time then fail, saying that our service was closed, giving our users the impression that there is a bug in our code.

The relevant line is [here](https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/lifecycle/internal/builder/multithreaded/MultiThreadedBuilder.java#L179). Instead of using `poll`, which simply returns when the parallel projects aren't finished yet, the code should shut down the executor and wait for it to complete the outstanding parallel work.

---

**Affects:** 3.5.4, 3.6.1

**Remote Links:**
- [GitHub Pull Request #272
](https://github.com/apache/maven/pull/272)

0 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Do you think one could add an IT  for that?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Isn't that the point that is shall **not** wait for others, but fail as fast as possible?

**Comment by jira-importer** [comment]
**[Stefan Oehme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oehme)** commented

It should fail fast, but not at the expense of correctness. As described above, Maven notifies us that the build has finished, but it's actually still executing things, which breaks the contract of these events. It's a totally uncoordinated shutdown, which I don't think is a good user experience.

We could try to make it wrap up faster, e.g. by calling `shutDownNow`, which will interrupt the execution threads. There's no hard guarantee that all Mojos support interrupts, but in general it should fail faster with that. Sounds good?

As for testing, it's actually pretty hard to reproduce with a small project, since it's a race condition between the goal failing and the VM shutting down. I may have to create a pretty large number of projects to make the test consistently fail. I'll see if that still gives a reasonable test execution time.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy), [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), WDYT? I tend to agree with [Stefan Oehme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oehme).

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I think fail-fast means: don't start any executing any new modules.
However, what to do with the still running modules? I think there are 2 valid solutions: let them finish or abort them. I prefer to first one, which should ensure all claimed resources are closed correctly. In case of the latter there should be a separate event like project_aborted.
Once all running modules are finished, the session can safely be finished and closed.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), this is how I understand Stefan's approach: cleanly shut down all open builders...

**Comment by jira-importer** [comment]
**[Stefan Oehme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oehme)** commented

Correct, my current solution waits for the concurrently running modules, but doesn't let any new ones start and then cleanly shuts down the build.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

In that case I like the idea, but I can't confirm that by code review. I would expect that the poll() is still required to empty the waiting projectSegments. I don't see any fork/join related code, but maybe I'm missing a detail.
Is it possible to create an integrationtest with a 3 modules: module A being longrunning,  module B being shortrunnig and failing, and module C shortrunning, but should not be called. It should confirm that A properly and that C is not called at all.

**Comment by jira-importer** [comment]
**[Stefan Oehme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oehme)** commented

> I would expect that the poll() is still required to empty the waiting projectSegments.

poll() doesn't wait, see its javadoc. It just returns null immediately if there is no finished project available.

> I don't see any fork/join related code, but maybe I'm missing a detail.

We wait for the executor to finish, which will wait for all scheduled work.

Yes, I can add a test.

**Comment by jira-importer** [comment]
**[Stefan Oehme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oehme)** commented

I have a test case ready, but after pulling in the latest master branch of the core-it project, everything is failing, even the "bootstrap" test, without any changes on my side. The first error message is:

```java
Failure to find org.apache.maven.plugins:maven-clean-plugin:jar:0.1-stub-SNAPSHOT in http://repo1.maven.org/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced
```

**Comment by jira-importer** [comment]
**[Stefan Oehme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oehme)** commented

Ah nevermind, it seems that the `-Dmaven.repo.local` argument does not tolerate relative paths.

**Comment by jira-importer** [comment]
**[Stefan Oehme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oehme)** commented

Test is here: https://github.com/apache/maven-integration-testing/pull/45. It fails on master and succeeds with my changes.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [0515cb1f0e258e1c22186a4e2dbdffa8db850b57](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=0515cb1f0e258e1c22186a4e2dbdffa8db850b57) and [3ccccc4cc3e4c7276496118e775dc68e631a3081](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git;a=commit;h=3ccccc4cc3e4c7276496118e775dc68e631a3081).
